src/toil/jobStores/abstractJobStore.py

```diff
@@ -114,6 +160,86 @@ def _checkJobStoreCreation(create, exists, jobStoreString):
             raise JobStoreCreationException("The job store '%s' does not exist, so there "
                                             "is nothing to restart." % jobStoreString)
 
+    def importFile(self, srcUrl):
+        """
+        Imports the file pointed at by sourceUrl into jobStore. The jobStoreFileId of the new
+        file is returned.
+
+        Currently Supported Schemes Include:
+            - 's3' for AWSJobStore
+                e.g. s3://bucket/key
+
+            - 'wasb' for AzureJobStore
+                e.g. wasb://container/blob
+
+            - 'file' for FileJobStore or other local files
+                e.g. file://local/file/path
+
+        :param str sourceUrl: URL that points to either a local file or to a file in a supported
+                jobStore that is currently installed.
+        :return The jobStoreFileId of imported file.
+        """
+        url = urlparse.urlparse(srcUrl)
+        with self.writeFileStream() as (writable, jobStoreFileID):
+            findJobStoreForUrl(url).readFromUrl(url, writable)
+            return jobStoreFileID
+
+    def exportFile(self, jobStoreFileID, dstUrl):
+        """
+        Exports file to destination pointed at by the destUrl.
+
+        :param str jobStoreFileID: The id of the file in the jobStore that should be exported.
+        :param str destUrl: URL that points to either a local file or to a file in a supported
+                jobStore e.g. 's3' for AWS.
+        """
+        url = urlparse.urlparse(dstUrl)
+        with self.readFileStream(jobStoreFileID) as readable:
+            findJobStoreForUrl(url).writeToUrl(readable, url)
+
+    @abstractstaticmethod
+    def readFromUrl(url, writable):
+        """
+        Reads the contents of the object at the specified location and writes it to the given
+        writable stream.
+
+        :param str url: URL pointing to a file in a jobStore
+        :param writable: a writable stream
+        """
+        raise NotImplementedError()
+
+    @abstractstaticmethod
+    def writeToUrl(readable, url):
+        """
+        Reads the contents of the given readable stream and writes it to the object at the
+        specified location.
+
+        :param str url: URL pointing to a file
+        :param readable: a readable stream
+        """
+        raise NotImplementedError()
+
+    @abstractclassmethod
+    def supportsUrl(cls, url):
+        """
+        Returns True if the jobStore supports the URL's scheme.
+
+        :param urlparse.ParseResult url: a parsed URL that may be supported
+        :return bool: returns true if the cls supports the URL
+        """
+        raise NotImplementedError()
+
+    @abstractstaticmethod
+    def _realizeUrl(url):
```

Comment:
Remove. This is an implementation detail that is not part of the API. You don't want to restrict subclasses to a specific way of implementing this detail unless you get something in return for paying that price.